### PR TITLE
Adding `IResult` implementation for FileStreamResult and LocalRedirectResult

### PR DESCRIPTION
**Pull request by uabarahona**
# Summary 
These changes add `IResult` implementation on `ActionResult`s that does not have it

# Details
The `ActionResult`s that are going to be worked on this PR are:

- [x] FileStreamResult
- [x] LocalRedirectResult

Task that needs to be done for this PR are:

- [x] Basic implementation for IResults
- [x] Add test suite to test the new IResults

Addresses part of #32565

### CODE REVIEW

File: src/Mvc/Mvc.Core/src/FileStreamResult.cs [hunk 1]
```diff
@@ -79,5 +81,42 @@ public override Task ExecuteResultAsync(ActionContext context)
...
         }
+
+        async Task IResult.ExecuteAsync(HttpContext httpContext)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            using (FileStream)
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/FileStreamResult.cs:92]: Out of curiosity, it is possible to share this implementation with the one in FileStreamExecutor (using an internal static method)? We could make the WriteFileAsync an optional parameter to the function to help with the overload.

`static ExecuteAsync(HttpContext context, ILogger logger, Func<..., Task>)`

Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/FileStreamResult.cs:92]: Same q with the redirect result.

Comment (barahonajm) [reply src/Mvc/Mvc.Core/src/FileStreamResult.cs:92]: That's a good idea!, I will make the change on night :)

Comment (barahonajm) [reply src/Mvc/Mvc.Core/src/FileStreamResult.cs]: @pranavkm I take me more than I thought to get back into this, I have applied your suggestions, would you take a look?

File: src/Mvc/Mvc.Core/src/FileStreamResult.cs [hunk 2]
```diff
@@ -79,5 +81,37 @@ public override Task ExecuteResultAsync(ActionContext context)
             var executor = context.HttpContext.RequestServices.GetRequiredService<IActionResultExecutor<FileStreamResult>>();
             return executor.ExecuteAsync(context, this);
         }
+
+        async Task IResult.ExecuteAsync(HttpContext httpContext)
+        {
+            var loggerFactory = httpContext.RequestServices.GetRequiredService<ILoggerFactory>();
+            var logger = loggerFactory.CreateLogger<RedirectResult>();
```
Comment (nil4) [inline src/Mvc/Mvc.Core/src/FileStreamResult.cs:88]: Is `RedirectResult` the relevant type parameter for this logger?

Comment (barahonajm) [reply src/Mvc/Mvc.Core/src/FileStreamResult.cs]: No, it should be `FileStreamResult`, thanks! already fixed

**Comment by barahonajm** [comment]
@pranavkm @nil4 I have fixed the logger types on all my other implementations, also if they way I implemented the approach suggested by @pranavkm is correct, I will migrate the other implementations either on this ticket or in a new PR later in this week.

Thanks!

File: src/Mvc/Mvc.Core/src/Infrastructure/FileStreamResultExecutor.cs [hunk 3]
```diff
@@ -26,6 +27,37 @@ public FileStreamResultExecutor(ILoggerFactory loggerFactory)
...
+            RangeItemHeaderValue? range,
+            long rangeLength)
+        {
+            return WriteFileAsyncInternal(context.HttpContext, result, range, rangeLength, Logger);
+        }
+
+        internal static async Task ExecuteAsyncInternal<TContext>(
+            TContext context,
+            FileStreamResult result,
+            Func<TContext, FileStreamResult, long?, bool, DateTimeOffset?, EntityTagHeaderValue?, (RangeItemHeaderValue?, long, bool)> SetHeadersAndLog,
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/Infrastructure/FileStreamResultExecutor.cs:58]:
```suggestion
            Func<TContext, FileStreamResult, long?, bool, DateTimeOffset?, EntityTagHeaderValue?, (RangeItemHeaderValue?, long, bool)> setHeadersAndLog,
```

File: src/Mvc/Mvc.Core/src/Infrastructure/FileStreamResultExecutor.cs [hunk 4]
```diff
@@ -26,6 +27,37 @@ public FileStreamResultExecutor(ILoggerFactory loggerFactory)
...
+            Func<TContext, FileStreamResult, RangeItemHeaderValue?, long, Task> WriteFileAsync,
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/Infrastructure/FileStreamResultExecutor.cs:59]:
```suggestion
            Func<TContext, FileStreamResult, RangeItemHeaderValue?, long, Task> writeFileAsync,
```

File: src/Mvc/Mvc.Core/src/FileStreamResult.cs [hunk 5]
```diff
@@ -79,5 +81,37 @@ public override Task ExecuteResultAsync(ActionContext context)
...
+
+        async Task IResult.ExecuteAsync(HttpContext httpContext)
+        {
+            var loggerFactory = httpContext.RequestServices.GetRequiredService<ILoggerFactory>();
+            var logger = loggerFactory.CreateLogger<FileStreamResult>();
+
+            Task writeFileAsync(HttpContext httpContext, FileStreamResult result, RangeItemHeaderValue? range, long rangeLength)
+                => FileStreamResultExecutor.WriteFileAsyncInternal(httpContext, this, range, rangeLength, logger!);
+
+            (RangeItemHeaderValue? range, long rangeLength, bool serveBody) setHeadersAndLog(
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/FileStreamResult.cs:93]: Can these be 
```suggestion
           static (RangeItemHeaderValue? range, long rangeLength, bool serveBody) setHeadersAndLog(
```

Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/FileStreamResult.cs:93]:
```suggestion
            (RangeItemHeaderValue? range, long rangeLength, bool serveBody) SetHeadersAndLog(
```

File: src/Mvc/Mvc.Core/src/FileStreamResult.cs [hunk 6]
```diff
@@ -79,5 +81,37 @@ public override Task ExecuteResultAsync(ActionContext context)
             var executor = context.HttpContext.RequestServices.GetRequiredService<IActionResultExecutor<FileStreamResult>>();
             return executor.ExecuteAsync(context, this);
         }
+
+        async Task IResult.ExecuteAsync(HttpContext httpContext)
+        {
+            var loggerFactory = httpContext.RequestServices.GetRequiredService<ILoggerFactory>();
+            var logger = loggerFactory.CreateLogger<FileStreamResult>();
+
+            Task writeFileAsync(HttpContext httpContext, FileStreamResult result, RangeItemHeaderValue? range, long rangeLength)
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/FileStreamResult.cs:90]:
```suggestion
            Task WriteFileAsync(HttpContext httpContext, FileStreamResult result, RangeItemHeaderValue? range, long rangeLength)
```

File: src/Mvc/Mvc.Core/src/FileStreamResult.cs [hunk 7]
```diff
@@ -79,5 +81,37 @@ public override Task ExecuteResultAsync(ActionContext context)
...
+                    EnableRangeProcessing,
+                    LastModified,
+                    EntityTag,
+                    logger!);
+
+            await FileStreamResultExecutor.ExecuteAsyncInternal(
+                httpContext,
+                this,
+                setHeadersAndLog,
+                writeFileAsync,
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/FileStreamResult.cs:113]:
```suggestion
                SetHeadersAndLog,
                WriteFileAsync,
```

File: src/Mvc/Mvc.Core/src/Infrastructure/LocalRedirectResultExecutor.cs [hunk 8]
```diff
@@ -45,36 +44,52 @@ public LocalRedirectResultExecutor(ILoggerFactory loggerFactory, IUrlHelperFacto
...
+                _logger);
+        }
+
+        internal static Task ExecuteAsyncInternal(
+            HttpContext httpContext,
+            LocalRedirectResult result,
+            Func<string, bool> isLocalUrl,
+            Func<string, string> getContent,
+            ILogger logger
+            )
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/Infrastructure/LocalRedirectResultExecutor.cs:63]:
```suggestion
            ILogger logger)
```

File: src/Mvc/Mvc.Core/src/Infrastructure/LocalRedirectResultExecutor.cs [hunk 9]
```diff
@@ -45,36 +44,52 @@ public LocalRedirectResultExecutor(ILoggerFactory loggerFactory, IUrlHelperFacto
         /// <inheritdoc />
         public virtual Task ExecuteAsync(ActionContext context, LocalRedirectResult result)
         {
-            if (context == null)
+            var urlHelper = result.UrlHelper ?? _urlHelperFactory.GetUrlHelper(context);
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/Infrastructure/LocalRedirectResultExecutor.cs:47]:
```suggestion
            if (context == null)
            {
                throw new ArgumentNullException(nameof(context));
            }
                
            var urlHelper = result.UrlHelper ?? _urlHelperFactory.GetUrlHelper(context);
```

File: src/Mvc/Mvc.Core/src/Infrastructure/LocalRedirectResultExecutor.cs [hunk 10]
```diff
@@ -45,36 +44,52 @@ public LocalRedirectResultExecutor(ILoggerFactory loggerFactory, IUrlHelperFacto
...
+
+            return ExecuteAsyncInternal(
+                context.HttpContext,
+                result,
+                urlHelper.IsLocalUrl,
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/Infrastructure/LocalRedirectResultExecutor.cs:52]: I feel a little iffy about this. This is now going to allocate a delegate per invocation. Could we have it fall-back in the implementation instead e.g.

```C#
internal static Task ExecuteAsyncInternal(
            HttpContext httpContext,
            LocalRedirectResult result,
            IUrlHelper? urlHelper)
{
  ...
   var isLocalUrl = urlHelper?.IsLocalUrl(url) ?? UrlHelperBase.CheckIsLocalUrl(url);
}

**Review by pranavkm — COMMENTED** [review]
Sorry for the delay in getting back

**Comment by pranavkm** [comment]
Thanks for your effort @barahonajm. We ended up deciding that it was best to avoid referencing MVC types from minimal actions especially for a common scenario such as the result types. We ended up duplicating the results and as such we no longer want to take this PR. Once again, thanks for your all your time!

**Comment by barahonajm** [comment]
@pranavkm that sound like a right decision, thanks for your guidance during this time. See you in another PR :)

**Pull request closed without merging**